src/test/lib/statistics/cardinality_estimator_test.cpp

```diff
@@ -1163,6 +1166,242 @@ TEST_F(CardinalityEstimatorTest, StatisticsCaching) {
   EXPECT_FALSE(statistics_cache->contains(predicate_node_2));
 }
 
+TEST_F(CardinalityEstimatorTest, StatisticsPruning) {
+  // clang-format off
+  const auto lqp =
+  PredicateNode::make(greater_than_(a_a, 50),
+    ProjectionNode::make(expression_vector(add_(a_a, 1), a_a, a_b, b_a, b_b),
+      JoinNode::make(JoinMode::Inner, equals_(a_b, b_a),
+        node_a,
+        node_b)));
+  // clang-format on
+
+  // Pruning without caching is not permitted.
+  EXPECT_THROW(estimator.prune_unused_statistics(), std::logic_error);
+  EXPECT_FALSE(estimator.cardinality_estimation_cache.statistics_by_lqp);
+  EXPECT_FALSE(estimator.cardinality_estimation_cache.required_column_expressions);
+
+  // Ensure required but pruned expressions are noticed.
+  if constexpr (HYRISE_DEBUG) {
+    // Because the cardinality estimator collects the required columns during estimation, we sneak in empty cached
+    // statistics. We trick the estimator to believe `node_a` was the entire LQP so it prunes all statistics and caches
+    // these completely pruned statistics for `node_a`. Any estimation that actually requires statistics should notice
+    // that statistics are missing.
+    estimator.guarantee_bottom_up_construction(node_a);
+    estimator.estimate_cardinality(node_a);
+    estimator.estimate_cardinality(node_b);
+    EXPECT_THROW(estimator.estimate_statistics(lqp), std::logic_error);
+    const auto projection = lqp->left_input();
+    lqp->set_left_input(node_a);
+    EXPECT_THROW(estimator.estimate_statistics(lqp), std::logic_error);
+    lqp->set_left_input(projection);
+    estimator.cardinality_estimation_cache.statistics_by_lqp.reset();
+    estimator.cardinality_estimation_cache.required_column_expressions.reset();
+  }
+
+  // Guaranteeing bottom-up construction (i.e., allowing caching) should enable statistics pruning and set the LQP in
+  // the cache. The required columns are not populated, yet.
+  estimator.guarantee_bottom_up_construction(lqp);
+  EXPECT_TRUE(estimator.cardinality_estimation_cache.statistics_by_lqp);
+  EXPECT_EQ(estimator.cardinality_estimation_cache.lqp, lqp);
+  ASSERT_TRUE(estimator.cardinality_estimation_cache.required_column_expressions);
+  ASSERT_TRUE(estimator.cardinality_estimation_cache.required_column_expressions->empty());
+
+  // Estimate with caching.
+  auto statistics = estimator.estimate_statistics(lqp);
+  ASSERT_EQ(statistics->column_statistics.size(), 5);
```

Comment:
When did we lose `b_b` here?
Can you add a comment why a_a, a_b, and b_a are valid but b_b not?

Reply:
Added a comment